frontend/ui/data/keyboardlayouts/vi_keyboard.lua

```diff
@@ -0,0 +1,76 @@
+-- Start with the english keyboard layout (deep copy, to not alter it)
+local vi_keyboard = require("util").tableDeepCopy(require("ui/data/keyboardlayouts/en_keyboard"))
+
+local IME = require("frontend/ui/data/keyboardlayouts/zh_ime")
+local util = require("util")
+
+-- see https://www.hieuthi.com/blog/2017/03/21/all-vietnamese-syllables.html
+local code_map = require("frontend/ui/data/keyboardlayouts/vi_telex_data")
+local ime = IME:new{
+    code_map = code_map,
+    partial_separators = {},
+    has_case = true,
+    exact_match = true,
+}
+
+local wrappedAddChars = function(inputbox, char)
+    local lowercase = char:lower()
+    ime:wrappedAddChars(inputbox, lowercase, char)
+end
+
+local function separate(inputbox)
+    ime:separate(inputbox)
+end
+
+local function wrappedDelChar(inputbox)
+    ime:wrappedDelChar(inputbox)
+end
+
+local function clear_stack()
+    ime:clear_stack()
+end
+
+local wrapInputBox = function(inputbox)
+    if inputbox._vi_wrapped == nil then
+        inputbox._vi_wrapped = true
+        local wrappers = {}
+
+        -- Wrap all of the navigation and non-single-character-input keys with
+        -- a callback to clear the tap window, but pass through to the
+        -- original function.
```

Comment:
I added a dedicated method for the Delete key in #9586 (`delNextChar`) ;).

Reply:
Good to know! Luckily we don't really care what's behind (next) when using `zh_ime` so no need to hook it :) Also I copied the comment from another keyboard file and it's not exactly accurate here, so I made a little update.